command/agent/cache/lease_cache.go

```diff
@@ -0,0 +1,743 @@
+	}
+
+	// Read the response from the index
+	resp, err := http.ReadResponse(bufio.NewReader(bytes.NewReader(index.Response)), nil)
+	if err != nil {
+		c.logger.Error("failed to deserialize response", "error", err)
+		return err
+	}
+
+	// Update the body in the reponse by the renewed secret
+	bodyBytes, err := json.Marshal(renewal.Secret)
+	if err != nil {
+		return err
+	}
+	resp.Body = ioutil.NopCloser(bytes.NewBuffer(bodyBytes))
+	resp.ContentLength = int64(len(bodyBytes))
+
+	// Serialize the response
+	var respBytes bytes.Buffer
+	err = resp.Write(&respBytes)
+	if err != nil {
+		c.logger.Error("failed to serialize updated response", "error", err)
+		return err
+	}
+
+	// Update the response in the index and set it in the cache
+	index.Response = respBytes.Bytes()
+	err = c.db.Set(index)
+	if err != nil {
+		c.logger.Error("failed to cache the proxied response", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+// computeIndexID results in a value that uniquely identifies a request
+// received by the agent. It does so by SHA256 hashing the serialized request
+// object containing the request path, query parameters and body parameters.
+func computeIndexID(req *SendRequest) (string, error) {
+	var b bytes.Buffer
+
+	// Serialze the request
+	if err := req.Request.Write(&b); err != nil {
+		return "", fmt.Errorf("failed to serialize request: %v", err)
+	}
+
+	// Reset the request body after it has been closed by Write
+	req.Request.Body = ioutil.NopCloser(bytes.NewBuffer(req.RequestBody))
+
+	// Append req.Token into the byte slice. This is needed since auto-auth'ed
+	// requests sets the token directly into SendRequest.Token
+	b.Write([]byte(req.Token))
+
+	sum := sha256.Sum256(b.Bytes())
+	return hex.EncodeToString(sum[:]), nil
+}
+
+// HandleCacheClear returns a handlerFunc that can perform cache clearing operations.
+func (c *LeaseCache) HandleCacheClear(ctx context.Context) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		req := new(cacheClearRequest)
+		if err := jsonutil.DecodeJSONFromReader(r.Body, req); err != nil {
+			if err == io.EOF {
+				err = errors.New("empty JSON provided")
+			}
+			respondError(w, http.StatusBadRequest, errwrap.Wrapf("failed to parse JSON input: {{err}}", err))
+			return
+		}
+
+		c.logger.Debug("received cache-clear request", "type", req.Type, "namespace", req.Namespace, "value", req.Value)
+
+		if err := c.handleCacheClear(ctx, req.Type, req.Namespace, req.Value); err != nil {
+			// Default to 500 on error, unless the user provided an invalid type,
+			// which would then be a 400.
+			httpStatus := http.StatusInternalServerError
+			if err == errInvalidType {
+				httpStatus = http.StatusBadRequest
+			}
+			respondError(w, httpStatus, errwrap.Wrapf("failed to clear cache: {{err}}", err))
+			return
+		}
+
+		return
+	})
+}
+
+func (c *LeaseCache) handleCacheClear(ctx context.Context, clearType string, clearValues ...interface{}) error {
```

Comment:
The `...interface{}` feels a little too flexible and easy to mess up, for instance if you are planning on clearing all you'd need to provide dummy values otherwise you will get caught in the below check. I wonder if it makes sense to provide a struct here instead. It would contain the Type, namespace, and Value? or instead of type it could be IndexType?

Reply:
Yeah I agree. I think this would be better as something like:
```go
type ClearInfo struct {
    RequestPaths []string
    Tokens []string
    TokenAccessors []string
    LeasePrefixes []string
    Everything bool
}
```

Reply:
I implemented the change using this struct:
```
type clearOptions struct {
	ClearType     string
	RequestPath   string
	Namespace     string
	Token         string
	TokenAccessor string
	Lease         string
}
```
I can probably get rid of `ClearType` since the types are technically present as fields, but how can I ensure that only one type is provided (with the exception of `RequestPath`, which can optionally include a non-empty `Namespace` value) without having a particular preference if more than one is provided. The ideal thing would be to not accept more than one. If `Token` and `Lease` where both provided, for instance, which one should handleCacheClear accept (again, the ideal thing would be to error out and say that only one field can be provided, but not sure how an "ensure only one field is set" can be coded without preference bias)?

Reply:
Why is it ideal to have only one? Why can't multiple be provided, and why can't multiple values of several of those fields be provided at once?